### PR TITLE
[ninegag] fix _VALID_URL regex and handle the use of other external providers

### PR DESCRIPTION
**Pull request by remitamine**

### CODE REVIEW

**Comment by dstftw** [comment]
Why not just extract `post_view['sourceUrl']`?

**Comment by remitamine** [comment]
this is right i will change it to get the url from sourceUrl and keep _EXTERNAL_VIDEO_PROVIDER to get the ie-key.

**Comment by dstftw** [comment]
No, just pass it without `ie_key` in order to delegate resolve to youtube-dl.

**Comment by remitamine** [comment]
if ie_key is known in advance. youtube-dl can access the right extractor directly.

**Comment by dstftw** [comment]
If this external video provider mapping changes on 9gag the whole extractions will be broken. In case of passing `sourceUrl` without `ie_key` it will work as usual.

**Comment by remitamine** [comment]
thanks for explanation,
i will remove _EXTERNAL_VIDEO_PROVIDER and use only `sourceUrl`.

**Comment by dstftw** [comment]
You can preserve `_EXTERNAL_VIDEO_PROVIDER` scenario as a fallback when `sourceUrl` is missing.

**Comment by remitamine** [comment]
i apply the requested changes.

File: youtube_dl/extractor/ninegag.py [hunk 1]
```diff
@@ -42,19 +37,44 @@ class NineGagIE(InfoExtractor):
...
 
         webpage = self._download_webpage(url, display_id)
 
         post_view = json.loads(self._html_search_regex(
             r'var postView = new app\.PostView\({\s*post:\s*({.+?}),\s*posts:\s*prefetchedCurrentPost', webpage, 'post view'))
 
-        youtube_id = post_view['videoExternalId']
+        ie_key = None
+        source_url = post_view.get('sourceUrl')
+        if not source_url or source_url == '':
```
Comment (dstftw) [inline youtube_dl/extractor/ninegag.py]: `not source_url` implies `source_url == ''`.

Comment (remitamine) [inline youtube_dl/extractor/ninegag.py]: i didn't know this. i will change it.

**Comment by remitamine** [comment]
i removed the unnecessary condition

**Comment by dstftw** [comment]
Thanks, cherry-picked.

**Pull request closed without merging**